### DISCUSSION TITLE
Why is my poll_next of tokio_stream::Stream only called once (returning pending)?

### DISCUSSION THREAD

**Discussion by i18nsite** [tokio-rs/tokio#6444] [Q&A]
As shown in the screenshot below, I am accessing my custom stream using while let Some(v) = stream.next().await {}. However, poll_next is only called once, returning pending, and it gets stuck.

<img width="1728" alt="image" src="https://github.com/tokio-rs/tokio/assets/145643935/08c0893e-af76-4537-9daf-33030b578258">

code is there https://github.com/i18n-site/rust/blob/main/map_await/src/lib.rs

**Comment by notgull** [comment]
https://github.com/i18n-site/rust/blob/02eb72ef365077d0a217142b7112f9dd3ecc3619/map_await/src/lib.rs#L36

You need to register the [`waker()`](https://doc.rust-lang.org/std/task/struct.Context.html#method.waker) of the [`Context`](https://doc.rust-lang.org/std/task/struct.Context.html) into something that will wake up the task later. For instance, for a channel, you need to [`wake`](https://doc.rust-lang.org/std/task/struct.Waker.html#method.wake) the waker when a new item is pushed into the channel.

**Comment by i18nsite** [comment]
![image](https://github.com/tokio-rs/tokio/assets/145643935/3fdac93e-7ab8-4aab-848d-24139cb3cb32)
thanks , I use  async_stream , it's easy

I have published this package here. I think it is quite convenient and practical. I use it to control the concurrency of S3 uploads.
https://crates.io/crates/map_await

https://crates.io/crates/s3_put
This is using map_unordered with s3_put to control uploads with a concurrency of 3
![image](https://github.com/tokio-rs/tokio/assets/145643935/9c633017-4619-48f2-87ec-687fb69a6f35)
